drivers: pwm: bbled: bbled-pwm driver to compatible with led api

Manimaran-A opened the pull request:
Zephyr has a PWM-LED driver which makes calls to a PWM driver to control the pin. The previous MCHP PWM-BBLED driver do not report its actual PWM cycles, instead it reported the clock source it used: 32KHz or 48MHz. This bug caused the LED-PWM driver set brightness to compute incorrect parameters passed to the PWM-BBLED driver. The BBLED hardware has a built-in divider of 256 therefore its maximum cycles are 128 and 48M/256 respectively. This change also simplified the calculations. A DT overlay for the MEC172x EVB was added to the PWM-LED sample to demonstrate driver operation. We found specifying >= 50 ms for the PWM period in the device tree nodes provided a good operating range for the LED-PWM driver parameter calculations. Note: BBLED hardware implements an 8-bit PWM with a fixed divider of 256 and a 12-bit frequency pre-scaler. Duty cycle is represented by an 8-bit value.

Manimaran-A commented:
@simonguinot @henrikbrixandersen Could you please help to review?

simonguinot commented:
> @simonguinot @henrikbrixandersen Could you please help to review?

Hi @Manimaran-A. I'll have a look at it this week-end.

simonguinot commented:
Hi @Manimaran-A. I apologize for the delay. Unfortunately I am not familiar enough with both the PWM subsystem and the MCHP PWM BBLED controller to help with this review.

I can see the division by 256 you added in the `get_cycles_per_sec` API function. And it looks right. But there are also other changes in the driver code. I don't understand them and the are not explained in the commit message. And I am not comfortable with approving changes that I don't fully understand. But I am sure that the PWM guys will be able to help with the review.
 
In addition I think you probably should move the addition of the DT overlay files into a dedicated commit.

albertofloyd commented:
@Manimaran-A  do split the commit in two: one for dts changes, one driver changes

albertofloyd commented on drivers/pwm/pwm_mchp_xec_bbled.c:
```diff
@@ -28,16 +28,15 @@ LOG_MODULE_REGISTER(pwmbbled_mchp_xec, CONFIG_PWM_LOG_LEVEL);
 #define XEC_PWM_BBLED_INPUT_FREQ_HI	48000000
 #define XEC_PWM_BBLED_INPUT_FREQ_LO	32768
 
-#define XEC_PWM_BBLED_MAX_FREQ_DIV	256U
```
Keep this and used it on new macros XEC_PWM_BBLED_MAX_PWM_FREQ_HI

Manimaran-A replied:
Ok.

Manimaran-A committed:
drivers: pwm: bbled: Update bbled-pwm, to be compatible with PWM-LED

Zephyr has a PWM-LED driver which makes calls to a PWM driver to control
the pin. The previous MCHP PWM-BBLED driver do not report its actual
PWM cycles, instead it reported the clock source it used: 32KHz or 48MHz.
This bug caused the LED-PWM driver set brightness to compute incorrect
parameters passed to the PWM-BBLED driver. The BBLED hardware has a
built-in divider of 256 therefore its maximum cycles are 128 and 48M/256
respectively. This change also simplified the calculations. A DT overlay
for the MEC172x EVB was added to the PWM-LED sample to demonstrate driver
operation. We found specifying >= 50 ms for the PWM period in the device
tree nodes provided a good operating range for the LED-PWM driver
parameter calculations.
Note: BBLED hardware implements an 8-bit PWM with a fixed divider of 256
and a 12-bit frequency pre-scaler. Duty cycle is represented by an 8-bit
value.

Signed-off-by: Manimaran A <manimaran.a@microchip.com>

Manimaran-A committed:
drivers: pwm: bbled: Update bbled-pwm dts files

Updated Overlay, dtsi and yaml files.

Signed-off-by: Manimaran A <manimaran.a@microchip.com>

MaureenHelm commented:
@jvasanth1 please take a look

simonguinot commented:
Hi @jvasanth1 and @Manimaran-A,

We need to merge this PR in order to remove the `led_mchp_xec` driver. The generic `led_pwm` driver should be used instead. Is that possible to get the PWM folks to review it ? @henrikbrixandersen, @anangl, can you help ?

jvasanth1 commented:
@albertofloyd can you help to review and approve?

fabiobaltieri commented on samples/drivers/led_pwm/boards/mec15xxevb_assy6853.overlay:
```diff
@@ -0,0 +1,81 @@
+
+&pinctrl {
+	led0_gpio156_invert: led0_gpio156_invert {
+		pinmux = < MCHP_XEC_PINMUX(0156, MCHP_AF1) >;
```
```suggestion
		pinmux = <MCHP_XEC_PINMUX(0156, MCHP_AF1)>;
```

Manimaran-A replied:
Ok, updated.

fabiobaltieri commented on samples/drivers/led_pwm/boards/mec15xxevb_assy6853.overlay:
```diff
@@ -0,0 +1,81 @@
+		microchip,output-func-invert;
+	};
+	led1_gpio157_invert: led1_gpio157_invert {
+		pinmux = < MCHP_XEC_PINMUX(0157, MCHP_AF1) >;
```
```suggestion
		pinmux = <MCHP_XEC_PINMUX(0157, MCHP_AF1)>;
```

Manimaran-A replied:
Updated.

fabiobaltieri commented on samples/drivers/led_pwm/boards/mec15xxevb_assy6853.overlay:
```diff
@@ -0,0 +1,81 @@
+		microchip,output-func-invert;
+	};
+	led2_gpio153_invert: led2_gpio153_invert {
+		pinmux = < MCHP_XEC_PINMUX(0153, MCHP_AF1) >;
```
```suggestion
		pinmux = <MCHP_XEC_PINMUX(0153, MCHP_AF1)>;
```

Manimaran-A replied:
Updated the comments.

fabiobaltieri commented on samples/drivers/led_pwm/boards/mec172xevb_assy6906.overlay:
```diff
@@ -0,0 +1,100 @@
+
+&pinctrl {
+	led0_gpio156_invert: led0_gpio156_invert {
+		pinmux = < MCHP_XEC_PINMUX(0156, MCHP_AF1) >;
```
```suggestion
		pinmux = <MCHP_XEC_PINMUX(0156, MCHP_AF1)>;
```

etc... you got the idea :-)

Manimaran-A replied:
Updated.
